File: sample_app/first_app/report/project_expense_report/project_expense_report.py
# Copyright (c) 2013, Leader Group and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from erpnext import get_company_currency, get_default_company


def execute(filters=None):
	columns, data = [], []
	columns = get_columns(filters)
	
	conditions, filters = get_conditions(filters)
	# get account list of Expense Account
	account_list = frappe.db.get_list('Account',
		filters = {
			'root_type': 'Expense',
			'is_group': 0
		},
		fields = ['name', 'account_number', 'account_name']
	)

	for account in account_list:
		amount = 0
		# get account total amount by debit amount minus credit amount from gl Entry
		gl_data = frappe.db.sql("""select account, (sum(debit) - sum(credit)) as amount
			from `tabGL Entry`
			where docstatus = 1 and is_cancelled != 1
				and account = '{0}' {1};""".format(account.get("name"),conditions),filters, as_dict=1)
		if gl_data:
			amount = gl_data[0].get("amount")
		
		if filters.get("show_zero_values"):
			# ignore amount only with none value
			if amount is not None:
				data.append({
					"account": account.get("name"),
					"account_number": account.get("account_number"),
					"account_name": account.get("account_name"),
					"amount": amount
				})
		else:
			# ignore amount with zero and none values
			if amount:
				data.append({
					"account": account.get("name"),
					"account_number": account.get("account_number"),
					"account_name": account.get("account_name"),
					"amount": amount
				})
	return columns, data

def get_columns(filters):
	columns = [
		{
			"fieldname": "account",
			"label": _("Account"),
			"fieldtype": "Link",
			"options": "Account",
			"width": 350
		},
		{
			"fieldname": "account_number",
			"label": _("Account #"),
			"fieldtype": "Data",
			"width": 120
		},
		{
			"fieldname": "account_name",
			"label": _("Account Name"),
			"fieldtype": "Data",
			"width": 300
		},
		{
			"fieldname": "amount",
			"label": _("Amount ({0})").format(get_company_currency(get_default_company())),
			"fieldtype": "Float",
			"width": 150
		}
	]
	return columns

def get_conditions(filters):
	conditions = ""
	if filters.get("from_date"):
		conditions += 'and posting_date >= %s' % frappe.db.escape(filters.get("from_date"), percent=False)
	if filters.get("to_date"):
		conditions += 'and posting_date <= %s' % frappe.db.escape(filters.get("to_date"), percent=False)
	if filters.get("project"):
		filters.project = frappe.parse_json(filters.get('project'))
		conditions += ' and project in %(project)s'
	return conditions, filters

File: leadergroup/leader_group/report/project_expense_report/project_expense_report.js
// Copyright (c) 2016, Leader Group and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.provide("leadergroup.project_expense_report");

leadergroup.project_expense_report = {
    "open_general_ledger": function(data) {
        if (!data.account) return;
        frappe.route_options = {
            "company": frappe.query_report.get_filter_value('company'),
            "from_date": frappe.query_report.get_filter_value('from_date'),
            "to_date": frappe.query_report.get_filter_value('to_date'),
            "project": frappe.query_report.get_filter_value('project')
        };
        account = data.account;
        frappe.route_options["account"] = account;
        frappe.set_route("query-report", "General Ledger");
    }
};

//only filter expense account
frappe.query_reports["Project Expense Report"] = {
	"filters": [
		{
            "fieldname": "company",
            "label": __("Company"),
            "fieldtype": "Link",
            "options": "Company",
            "default": frappe.defaults.get_user_default("Company") || frappe.defaults.get_global_default("Company"),
            "reqd": 1,
            "hidden": 1
        },
		{
			"fieldname":"project",
			"label": __("Project"),
			"fieldtype": "MultiSelectList",
			get_data: function(txt) {
				return frappe.db.get_link_options('Project', txt, {
					company: frappe.query_report.get_filter_value("company")
				});
			}
		},
		{
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
            "reqd": 1
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.get_today(),
            "reqd": 1
        },
		{
			"fieldname": "show_zero_values",
			"label": __("Show Zero Values"),
			"fieldtype": "Check"
		},
        {
            "fieldname": "printed_on",
            "label": __("Printed On"),
            "fieldtype": "Data",
            "default": moment(frappe.datetime.now_datetime()).format("MM-DD-YYYY HH:mm:ss"),
            "read_only": 1
        }
	],
    "formatter": function(value, row, column, data, default_formatter) {
        if (data && column.fieldname=="account_name") {
            value = data.account_name || value;

            column.link_onclick = "leadergroup.project_expense_report.open_general_ledger(" + JSON.stringify(data) + ")";
        }
        value = default_formatter(value, row, column, data);
        return value;
    },
    
    onload: function(report) {
        const views_menu = report.page.add_custom_button_group(__('General Ledger Report'));

        report.page.add_custom_menu_item(views_menu, __("General Ledger"), function() {
            var filters = report.get_values();
            frappe.set_route('query-report', 'General Ledger', filters);
        });
    }
};